File: backend/app/schemas/workflow_generation.py
"""
Workflow Generation Pydantic Schemas

Defines request/response schemas for workflow generation API endpoints.
"""
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime
from enum import Enum


class WorkflowStatus(str, Enum):
    """Status of generated workflow"""
    DRAFT = "draft"
    ACTIVE = "active"
    ARCHIVED = "archived"


class ExecutionStatus(str, Enum):
    """Status of workflow execution"""
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


class WorkflowCategory(str, Enum):
    """Categories for workflow templates"""
    DATA_PROCESSING = "data_processing"
    API_INTEGRATION = "api_integration"
    AUTOMATION = "automation"
    ETL = "etl"
    MACHINE_LEARNING = "machine_learning"
    DOCUMENT_PROCESSING = "document_processing"
    NOTIFICATION = "notification"
    SCHEDULING = "scheduling"
    CUSTOM = "custom"


# ============== WorkflowTemplate Schemas ==============

class WorkflowTemplateBase(BaseModel):
    """Base schema for WorkflowTemplate"""
    name: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = None
    category: WorkflowCategory = WorkflowCategory.CUSTOM


class WorkflowTemplateCreate(WorkflowTemplateBase):
    """Schema for creating a new WorkflowTemplate"""
    template_schema: Dict[str, Any] = Field(..., description="Workflow structure definition")
    parameters: Optional[Dict[str, Any]] = Field(None, description="Input parameters schema")
    is_public: bool = Field(False, description="Whether the template is public")


class WorkflowTemplateUpdate(BaseModel):
    """Schema for updating a WorkflowTemplate"""
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = None
    category: Optional[WorkflowCategory] = None
    template_schema: Optional[Dict[str, Any]] = None
    parameters: Optional[Dict[str, Any]] = None
    is_public: Optional[bool] = None


class WorkflowTemplateResponse(WorkflowTemplateBase):
    """Schema for WorkflowTemplate response"""
    id: int
    template_schema: Dict[str, Any]
    parameters: Optional[Dict[str, Any]] = None
    is_public: bool
    user_id: Optional[int] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class WorkflowTemplateListResponse(BaseModel):
    """Schema for paginated template list"""
    templates: List[WorkflowTemplateResponse]
    total: int
    limit: int
    offset: int


# ============== GeneratedWorkflow Schemas ==============

class GeneratedWorkflowBase(BaseModel):
    """Base schema for GeneratedWorkflow"""
    name: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = None


class GeneratedWorkflowCreate(GeneratedWorkflowBase):
    """Schema for creating a new GeneratedWorkflow"""
    template_id: Optional[int] = None
    workflow_schema: Dict[str, Any] = Field(..., description="Generated workflow structure")
    generation_params: Optional[Dict[str, Any]] = Field(None, description="Parameters used for generation")
    status: WorkflowStatus = WorkflowStatus.DRAFT


class GeneratedWorkflowResponse(GeneratedWorkflowBase):
    """Schema for GeneratedWorkflow response"""
    id: int
    template_id: Optional[int] = None
    user_id: int
    workflow_schema: Dict[str, Any]
    generation_params: Optional[Dict[str, Any]] = None
    status: WorkflowStatus
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class GeneratedWorkflowListResponse(BaseModel):
    """Schema for paginated generated workflow list"""
    workflows: List[GeneratedWorkflowResponse]
    total: int
    limit: int
    offset: int


# ============== WorkflowExecution Schemas ==============

class WorkflowExecutionBase(BaseModel):
    """Base schema for WorkflowExecution"""
    pass


class WorkflowExecutionCreate(WorkflowExecutionBase):
    """Schema for creating a new WorkflowExecution"""
    generated_workflow_id: int
    input_data: Optional[Dict[str, Any]] = None


class WorkflowExecutionResponse(WorkflowExecutionBase):
    """Schema for WorkflowExecution response"""
    id: int
    generated_workflow_id: int
    status: ExecutionStatus
    input_data: Optional[Dict[str, Any]] = None
    output_data: Optional[Dict[str, Any]] = None
    execution_time_ms: Optional[float] = None
    error_message: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


# ============== WorkflowPattern Schemas ==============

class WorkflowPatternBase(BaseModel):
    """Base schema for WorkflowPattern"""
    name: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = None


class WorkflowPatternResponse(WorkflowPatternBase):
    """Schema for WorkflowPattern response"""
    id: int
    pattern_schema: Dict[str, Any]
    usage_count: int
    success_rate: float
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class WorkflowPatternListResponse(BaseModel):
    """Schema for pattern list"""
    patterns: List[WorkflowPatternResponse]


# ============== Workflow Generation Request/Response ==============

class WorkflowGenerationRequest(BaseModel):
    """Schema for AI-powered workflow generation request"""
    description: str = Field(..., description="Natural language description of the desired workflow")
    parameters: Optional[Dict[str, Any]] = Field(None, description="Input parameters for the workflow")
    category: Optional[WorkflowCategory] = Field(None, description="Optional category hint")


class WorkflowGenerationResponse(BaseModel):
    """Schema for AI-powered workflow generation response"""
    workflow: GeneratedWorkflowResponse
    matched_template: Optional[WorkflowTemplateResponse] = None
    pattern_matches: List[WorkflowPatternResponse] = []
    optimization_suggestions: List[str] = []


class WorkflowExecutionRequest(BaseModel):
    """Schema for executing a generated workflow"""
    workflow_id: int
    input_data: Optional[Dict[str, Any]] = None


class PatternRecognitionRequest(BaseModel):
    """Schema for pattern recognition request"""
    workflow_schema: Dict[str, Any]


class PatternRecognitionResponse(BaseModel):
    """Schema for pattern recognition response"""
    matched_pattern: Optional[WorkflowPatternResponse] = None
    confidence: float = 0.0
    similar_patterns: List[WorkflowPatternResponse] = []


# ============== Workflow Optimization ==============

class WorkflowOptimizationRequest(BaseModel):
    """Schema for workflow optimization request"""
    workflow_schema: Dict[str, Any]


class WorkflowOptimizationResponse(BaseModel):
    """Schema for workflow optimization response"""
    optimized_schema: Dict[str, Any]
    improvements: List[str]
    estimated_performance_gain: float = 0.0

File: frontend/src/types/workflowGeneration.ts
/**
 * Workflow Generation Types
 * 
 * TypeScript interfaces matching the backend Pydantic schemas
 * for the Automated Workflow Generation System.
 */

// Enum Types
export type WorkflowStatus = 'draft' | 'active' | 'archived';
export type ExecutionStatus = 'pending' | 'running' | 'completed' | 'failed';
export type WorkflowCategory = 
  | 'data_processing' 
  | 'api_integration' 
  | 'automation' 
  | 'etl' 
  | 'machine_learning' 
  | 'document_processing' 
  | 'notification' 
  | 'scheduling' 
  | 'custom';

// Base Types
export interface WorkflowTemplateBase {
  name: string;
  description?: string;
  category: WorkflowCategory;
}

export interface GeneratedWorkflowBase {
  name: string;
  description?: string;
}

export interface WorkflowPatternBase {
  name: string;
  description?: string;
}

// Workflow Template
export interface WorkflowTemplate extends WorkflowTemplateBase {
  id: number;
  template_schema: WorkflowSchema;
  parameters?: Record<string, any>;
  is_public: boolean;
  user_id?: number;
  created_at: string;
  updated_at: string;
}

export interface WorkflowTemplateCreate extends WorkflowTemplateBase {
  template_schema: WorkflowSchema;
  parameters?: Record<string, any>;
  is_public?: boolean;
}

export interface WorkflowTemplateUpdate {
  name?: string;
  description?: string;
  category?: WorkflowCategory;
  template_schema?: WorkflowSchema;
  parameters?: Record<string, any>;
  is_public?: boolean;
}

export interface WorkflowTemplateListResponse {
  templates: WorkflowTemplate[];
  total: number;
  limit: number;
  offset: number;
}

// Generated Workflow
export interface GeneratedWorkflow extends GeneratedWorkflowBase {
  id: number;
  template_id?: number;
  user_id: number;
  workflow_schema: WorkflowSchema;
  generation_params?: Record<string, any>;
  status: WorkflowStatus;
  created_at: string;
  updated_at: string;
}

export interface GeneratedWorkflowCreate extends GeneratedWorkflowBase {
  template_id?: number;
  workflow_schema: WorkflowSchema;
  generation_params?: Record<string, any>;
  status?: WorkflowStatus;
}

export interface GeneratedWorkflowListResponse {
  workflows: GeneratedWorkflow[];
  total: number;
  limit: number;
  offset: number;
}

// Workflow Execution
export interface WorkflowExecution {
  id: number;
  generated_workflow_id: number;
  status: ExecutionStatus;
  input_data?: Record<string, any>;
  output_data?: Record<string, any>;
  execution_time_ms?: number;
  error_message?: string;
  created_at: string;
}

export interface WorkflowExecutionCreate {
  generated_workflow_id: number;
  input_data?: Record<string, any>;
}

export interface WorkflowExecutionRequest {
  workflow_id: number;
  agent_id: number;
  input_data?: Record<string, any>;
}

export interface WorkflowSchemaExecutionRequest {
  agent_id: number;
  workflow_schema: WorkflowSchema;
  input_data?: Record<string, any>;
}

// Workflow Pattern
export interface WorkflowPattern extends WorkflowPatternBase {
  id: number;
  pattern_schema: Record<string, any>;
  usage_count: number;
  success_rate: number;
  created_at: string;
  updated_at: string;
}

export interface WorkflowPatternListResponse {
  patterns: WorkflowPattern[];
}

// Workflow Schema
export interface WorkflowSchema {
  name: string;
  description?: string;
  category?: WorkflowCategory;
  complexity?: 'simple' | 'moderate' | 'complex';
  steps: WorkflowStep[];
  triggers?: WorkflowTrigger[];
  metadata?: Record<string, any>;
  _optimizations?: {
    applied_at?: string;
    improvements?: string[];
  };
}

export interface WorkflowStep {
  name: string;
  type: string;
  description?: string;
  inputs?: Record<string, any>;
  outputs?: Record<string, any>;
  depends_on?: string[];
  config?: Record<string, any>;
}

export interface WorkflowTrigger {
  type: 'manual' | 'schedule' | 'webhook' | 'event';
  config?: Record<string, any>;
}

// Generation Request/Response
export interface WorkflowGenerationRequest {
  description: string;
  parameters?: Record<string, any>;
  category?: WorkflowCategory;
}

export interface WorkflowGenerationResponse {
  workflow: GeneratedWorkflow;
  matched_template?: WorkflowTemplate;
  pattern_matches: WorkflowPattern[];
  optimization_suggestions: string[];
}

// Pattern Recognition
export interface PatternRecognitionRequest {
  workflow_schema: WorkflowSchema;
}

export interface PatternRecognitionResponse {
  matched_pattern?: WorkflowPattern;
  confidence: number;
  similar_patterns: WorkflowPattern[];
}

// Optimization
export interface WorkflowOptimizationRequest {
  workflow_schema: WorkflowSchema;
}

export interface WorkflowOptimizationResponse {
  optimized_schema: WorkflowSchema;
  improvements: string[];
  estimated_performance_gain: number;
}

// Workflow Builder State
export interface WorkflowBuilderState {
  currentSchema: WorkflowSchema | null;
  selectedStep: string | null;
  isDirty: boolean;
  isLoading: boolean;
  error: string | null;
}

// AI Generator State
export interface AIWorkflowGeneratorState {
  description: string;
  parameters: Record<string, any>;
  isGenerating: boolean;
  generatedWorkflow: GeneratedWorkflow | null;
  suggestions: string[];
  error: string | null;
}

// Pattern Visualizer State
export interface PatternVisualizerState {
  patterns: WorkflowPattern[];
  selectedPattern: WorkflowPattern | null;
  isLoading: boolean;
}
